### PR TITLE
dexterstorey/multi user

### PR DESCRIPTION
**Pull request by DexterStorey**
Based on feedback from Cal Team, the Cal AI Bot should be able to schedule meetings between 2 Cal users in their availability ranges.

### CODE REVIEW

**Comment by DexterStorey** [comment]
Still workshopping this a bit

**Comment by DexterStorey** [comment]
We will likely call this a v1.1 feature and wait until the [v1.0 PR](https://github.com/calcom/cal.com/pull/10992) is merged to implement this

File: apps/ai/src/app/api/receive/route.ts [hunk 1]
```diff
@@ -53,7 +54,7 @@ export const POST = async (request: NextRequest) => {
         },
       },
     },
-    where: { email: envelope.from },
+    where: { email: envelope.from, credentials: { some: { appId: env.APP_ID } } },
```
Comment (tedspare) [inline apps/ai/src/app/api/receive/route.ts:57]: What does this do? Only include the user's credentials for our installation?

File: apps/ai/src/app/api/agent/route.ts [hunk 2]
```diff
@@ -43,7 +43,7 @@ export const POST = async (request: NextRequest) => {
     return new NextResponse("ok");
   } catch (error) {
     return new NextResponse(
-      error.message || "Something went wrong. Please try again or reach out for help.",
+      (error as Error).message || "Something went wrong. Please try again or reach out for help.",
```
Comment (tedspare) [inline apps/ai/src/app/api/agent/route.ts:46]: `... catch (error: Error) ...` instead of `as Error`

File: apps/ai/src/app/api/receive/route.ts [hunk 3]
```diff
@@ -90,12 +91,13 @@ export const POST = async (request: NextRequest) => {
...
 
   // Pre-fetch data relevant to most bookings.
-  const [eventTypes, availability] = await Promise.all([
+  const [eventTypes, availability, users] = await Promise.all([
     fetchEventTypes(),
     fetchAvailability({
       dateFrom: now,
       dateTo: now,
     }),
+    extractUsers(`${parsed.text} ${parsed.subject}`),
```
Comment (tedspare) [inline apps/ai/src/app/api/receive/route.ts:100]: Should subject be before text? Also, will users have to mention their/others' Cal usernames in emails for this to work? What about using original `to` email instead?

File: apps/ai/src/tools/getAvailability.ts [hunk 4]
```diff
@@ -63,6 +66,12 @@ const getAvailabilityTool = new DynamicStructuredTool({
   },
   name: "getAvailability",
   schema: z.object({
+    userId: z
+      .string()
+      .optional()
+      .describe(
+        "The user ID of an external user to fetch availability for. If not provided, uses the current user."
```
Comment (tedspare) [inline apps/ai/src/tools/getAvailability.ts:73]: Super nice!

File: apps/ai/src/types/user.ts [hunk 5]
```diff
@@ -2,12 +2,10 @@ import type { EventType } from "./eventType";
...
-  userIdIV: string;
   email: string;
   timeZone: string;
-  apiKeyHashed: string;
-  apiKeyIV: string;
   eventTypes: EventType[];
   workingHours: WorkingHours[];
 };
+
+export type UserList = { id: number; email?: string; username?: string }[];
```
Comment (tedspare) [inline apps/ai/src/types/user.ts:11]: `export type UserList = Array<User & { id: number; }>`

File: apps/ai/src/utils/extractUsers.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            email: true,
+          },
+          where: {
+            username: {
+              in: usernames,
+            },
+          },
+        })
+      ).map((user, n) => ({ username: usernames[n] as string, id: user.id, email: user.email }))
+    : [];
```
Comment (tedspare) [inline apps/ai/src/utils/extractUsers.ts:24]: Nice!

**Review by tedspare — APPROVED** [review]
LGTM! Only nitpicks:
- Contains commits from main (can separate these in the future)
- Could break up some long chains into multiple vars for readability

Great work

**Pull request closed without merging**